ncm-download/src/main/perl/download.pm

```diff
@@ -301,19 +320,28 @@ sub retrieve
     if ($timestamp_remote) {
         if ($timestamp_remote > $timestamp_existing) { # If local file doesn't exist, this still works
             if ($timestamp_remote <= $timestamp_threshold) { # Also prevents future files
+                $self->_cleanup_tmpfile($tmpfn);
                 my $proc = CAF::Process->new(\@cmd, stderr => \my $errs, log => $self);
                 $proc->execute();
                 if (!POSIX::WIFEXITED($?) || POSIX::WEXITSTATUS($?) != 0) {
                     # Only warn as there may be attempts from multiple proxies.
                     $self->warn("curl failed (". ($?>>8) ."): $errs\nCommand = $proc");
+                    $self->_cleanup_tmpfile($tmpfn);
                     return 0; # Curl barfed, so we return.
                 } else {
                     if ($CAF::Object::NoAction) {
                         $self->debug(1, "We have sucessfully pretended to retrieve a new copy of the file $fn");
                     } else {
-                        $self->debug(1, "We seem to have sucessfully retrieved a new copy of the file $fn");
+                        $self->debug(1, "We seem to have sucessfully retrieved a new copy of the file $fn to $tmpfn");
+                    }
+                    if ($self->move($tmpfn, $fn)) {
```

Comment:
we should document here (and possibly in CAF::Path) that `$self->move` use `FIle::Copy::move` which uses a `rename`, which is atomic within the same filesystem (which is the case for file in the same directory) so this does what we want safely